File: 1022. Sum of Root To Leaf Binary Numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int dfs(TreeNode* root, int curr) {
        if(root == NULL)
            return 0;

        // form binary number
        curr = (curr << 1) | root->val;

        // if leaf → return number
        if(root->left == NULL && root->right == NULL)
            return curr;

        // sum of left and right
        return dfs(root->left, curr) + dfs(root->right, curr);
    }
    int sumRootToLeaf(TreeNode* root) {
        return dfs(root, 0);
        
    }
};

File: 1022. Sum of Root To Leaf Binary Numbers.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function sumRootToLeaf(root: TreeNode | null): number {
    function dfs(node: TreeNode | null, curr: number): number {
        if (node === null) return 0;

        // build binary number
        curr = (curr << 1) | node.val;

        // leaf node
        if (node.left === null && node.right === null) {
            return curr;
        }

        return dfs(node.left, curr) + dfs(node.right, curr);
    }

    return dfs(root, 0);
};